src/onnx/parse_mean.cpp

```diff
@@ -0,0 +1,107 @@
+            lens = new_lens;
+        }
+    });
+
+    // Compute multidirectional broadcasted shape compatible with all inputs
+    std::vector<std::size_t> output_lens(max_size, 1);
+    for(std::size_t i = 0; i < max_size; ++i)
+    {
+        std::for_each(input_lens.begin(), input_lens.end(), [&](const auto& lens) {
+            if(lens[i] != 1)
+            {
+                if(output_lens[i] == 1)
+                    output_lens[i] = lens[i];
+                else if(output_lens[i] != lens[i])
+                    MIGRAPHX_THROW("Parse Mean: Shapes are incompatible for multi-broadcast");
+            }
+        });
+    }
+
+    return output_lens;
+}
+
+struct parse_mean : op_parser<parse_mean>
+{
+    std::vector<op_desc> operators() const { return {{"Mean"}}; }
+
+    /// Calculates the element-wise mean of n>=1 input tensors
+    instruction_ref parse(const op_desc& /*opd*/,
+                          const onnx_parser& /*parser*/,
+                          const onnx_parser::node_info& info,
+                          std::vector<instruction_ref> args) const
+    {
+        auto num_data = args.size();
+        if(num_data == 1)
+            return args[0];
+
+        std::vector<std::vector<std::size_t>> input_lens;
+        std::transform(args.begin(),
+                       args.end(),
+                       std::back_inserter(input_lens),
+                       [&](const auto arg) { return arg->get_shape().lens(); });
+        auto output_lens = compute_output_lens(input_lens);
+        auto output_type = args[0]->get_shape().type();
+
+        auto mean    = args[0];
+        auto divisor = info.add_literal(
+            migraphx::literal{migraphx::shape{mean->get_shape().type()}, {num_data}});
+        divisor = info.add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", output_lens}}), divisor);
+        if(not std::equal(mean->get_shape().lens().begin(),
+                          mean->get_shape().lens().end(),
+                          output_lens.begin(),
+                          output_lens.end()))
+            mean = info.add_instruction(
+                migraphx::make_op("multibroadcast", {{"out_lens", output_lens}}), mean);
+
+        // Pre-divide each tensor element-wise by n to reduce risk of overflow during summation
+        mean = info.add_instruction(migraphx::make_op("div"), mean, divisor);
+
+        std::for_each(args.begin() + 1, args.end(), [&](auto& data_i) {
+            if(data_i->get_shape().type() != output_type)
+                MIGRAPHX_THROW("Parse Mean: All inputs must have the same data type.");
+
+            auto data_i_lens = data_i->get_shape().lens();
+            if(not std::equal(
+                   data_i_lens.begin(), data_i_lens.end(), output_lens.begin(), output_lens.end()))
+                data_i = info.add_instruction(
+                    migraphx::make_op("multibroadcast", {{"out_lens", output_lens}}), data_i);
```

Comment:
Can we use `add_common_op` to insert the broadcasts?

Reply:
Actually, `info.add_broadcastable_binary_op` would work as well.

Reply:
Changed